File: yandex_two_sets.cpp
#include <iostream>
#include <vector>
#include <numeric>
#include <random>
#include <unordered_map>
#include <algorithm>

using namespace std;

typedef long long ll;

void solve() {
    int n = 40;
    vector<ll> a(n);
    for (int i = 0; i < n; ++i) {
        if (!(cin >> a[i])) return;
    }

    // Храним {сумма: маска_подмножества}
    // Маска — это 64-битное число, где i-й бит означает включение i-го элемента
    unordered_map<ll, ll> sums;
    
    // Инициализация генератора случайных чисел
    mt19937_64 rng(1337); 
    
    while (true) {
        ll mask = 0;
        ll current_sum = 0;
        
        // Генерируем случайную маску для 40 элементов
        // Каждое подмножество равновероятно
        for (int i = 0; i < n; ++i) {
            if (rng() % 2) {
                mask |= (1LL << i);
                current_sum += a[i];
            }
        }

        if (mask == 0) continue;

        if (sums.count(current_sum)) {
            ll prev_mask = sums[current_sum];
            if (prev_mask == mask) continue;

            // Убираем общие элементы (пересечение подмножеств)
            ll common = prev_mask & mask;
            ll final_mask1 = prev_mask ^ common;
            ll final_mask2 = mask ^ common;

            // Если после удаления общих элементов одно из множеств стало пустым
            if (final_mask1 == 0 || final_mask2 == 0) continue;

            // Вывод первого множества
            vector<int> res1, res2;
            for (int i = 0; i < n; ++i) {
                if ((final_mask1 >> i) & 1) res1.push_back(i + 1);
                if ((final_mask2 >> i) & 1) res2.push_back(i + 1);
            }

            cout << res1.size() << "\n";
            for (int i = 0; i < res1.size(); ++i) {
                cout << res1[i] << (i == res1.size() - 1 ? "" : " ");
            }
            cout << "\n";

            // Вывод второго множества
            cout << res2.size() << "\n";
            for (int i = 0; i < res2.size(); ++i) {
                cout << res2[i] << (i == res2.size() - 1 ? "" : " ");
            }
            cout << "\n";

            return;
        }

        sums[current_sum] = mask;
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    solve();
    return 0;
}

File: yandex_two_sets.py
import sys
import random

def solve():
    # Читаем все числа и преобразуем в список
    data = sys.stdin.read().split()
    if not data: return
    a = [int(x) for x in data]
    n = len(a)
    
    # Храним {сумма: кортеж_индексов}
    sums = {}
    
    # Чтобы ускорить процесс, будем брать случайные подмножества 
    # фиксированного размера (например, 20 элементов из 40)
    # Этого более чем достаточно для нахождения коллизии
    indices_pool = list(range(n))
    
    while True:
        # random.sample работает быстрее, чем ручной цикл по 40 элементам
        sample_idx = random.sample(indices_pool, 20)
        current_sum = sum(a[i] for i in sample_idx)
        
        if current_sum in sums:
            prev_indices = sums[current_sum]
            
            # Превращаем в множества для поиска пересечений
            s1 = set(sample_idx)
            s2 = set(prev_indices)
            
            # Если наборы идентичны (маловероятно), продолжаем
            if s1 == s2: continue
            
            # Убираем общие индексы
            common = s1 & s2
            res1 = [i + 1 for i in s1 if i not in common]
            res2 = [i + 1 for i in s2 if i not in common]
            
            # Вывод по формату
            print(len(res1))
            print(*(res1))
            print(len(res2))
            print(*(res2))
            return
        
        sums[current_sum] = tuple(sample_idx)

if __name__ == '__main__':
    solve()
